akka-stream-tests/src/test/scala/akka/stream/io/TlsSpec.scala

```diff
@@ -0,0 +1,312 @@
+    keyStore.load(getClass.getResourceAsStream("/keystore"), password.toCharArray)
+
+    val trustStore = KeyStore.getInstance(KeyStore.getDefaultType)
+    trustStore.load(getClass.getResourceAsStream("/truststore"), password.toCharArray)
+
+    val keyManagerFactory = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm)
+    keyManagerFactory.init(keyStore, password.toCharArray)
+
+    val trustManagerFactory = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm)
+    trustManagerFactory.init(trustStore)
+
+    val context = SSLContext.getInstance("TLS")
+    context.init(keyManagerFactory.getKeyManagers, trustManagerFactory.getTrustManagers, new SecureRandom)
+    context
+  }
+
+  class Timeout()(implicit system: ActorSystem) extends AsyncStage[ByteString, ByteString, Unit] {
+    private var last: ByteString = _
+
+    override def initAsyncInput(ctx: AsyncContext[ByteString, Unit]) = {
+      val cb = ctx.getAsyncCallback()
+      system.scheduler.scheduleOnce(2.seconds)(cb.invoke(()))(system.dispatcher)
+    }
+
+    override def onAsyncInput(u: Unit, ctx: AsyncContext[ByteString, Unit]) =
+      ctx.fail(new TimeoutException(s"timeout expired, last element was $last"))
+
+    override def onPush(elem: ByteString, ctx: AsyncContext[ByteString, Unit]) = {
+      last = elem
+      if (ctx.isHoldingDownstream) ctx.pushAndPull(elem)
+      else ctx.holdUpstream()
+    }
+
+    override def onPull(ctx: AsyncContext[ByteString, Unit]) =
+      if (ctx.isFinishing) ctx.pushAndFinish(last)
+      else if (ctx.isHoldingUpstream) ctx.pushAndPull(last)
+      else ctx.holdDownstream()
+
+    override def onUpstreamFinish(ctx: AsyncContext[ByteString, Unit]) =
+      if (ctx.isHoldingUpstream) ctx.absorbTermination()
+      else ctx.finish()
+  }
+}
+
+class TlsSpec extends AkkaSpec("akka.loglevel=INFO\nakka.actor.debug.receive=off") {
+  import TlsSpec._
+
+  import system.dispatcher
+  implicit val materializer = ActorFlowMaterializer()
+
+  import FlowGraph.Implicits._
+
+  "StreamTLS" must {
+    import StreamTls._
+
+    // All of these are reusable
+    val sslContext = initSslContext()
+
+    val debug = Flow[SslTlsInbound].map { x ⇒
+      x match {
+        case SessionTruncated   ⇒ system.log.debug(s" ----------- truncated ")
+        case SessionBytes(_, b) ⇒ system.log.debug(s" ----------- (${b.size}) ${b.take(32).utf8String}")
+      }
+      x
+    }
+
+    val cipherSuites = NegotiateNewSession.withCipherSuites("TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA", "TLS_RSA_WITH_AES_128_CBC_SHA")
+    def clientTls(closing: Closing) = StreamTls(sslContext, cipherSuites, Client, closing)
+    def serverTls(closing: Closing) = StreamTls(sslContext, cipherSuites, Server, closing)
+
+    trait Named {
+      def name: String = getClass.getName.reverse.dropWhile(c ⇒ "$0123456789".indexOf(c) != -1).takeWhile(_ != '$').reverse
+    }
+
+    trait CommunicationSetup extends Named {
```

Comment:
add a comment to explaint that these traits are meant to be composed together, and how it works in a nutshell